mono/metadata/marshal-ilgen.c

```diff
@@ -2700,11 +2700,32 @@ emit_marshal_array_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 		if (m_class_is_blittable (eklass)) {
 			/* free memory allocated (if any) by MONO_MARSHAL_CONV_ARRAY_LPARRAY */
 
+			int param_num = spec ? spec->data.array_data.param_num : -1;
+			int num_elem = spec ? spec->data.array_data.num_elem : 0;
+			if (spec && spec->data.array_data.elem_mult == 0)
+				/* param_num is not specified */
+				param_num = -1;
+
+			if (param_num == -1 && num_elem <= 0)
+				num_elem = 0;
+
 			mono_mb_emit_ldarg (mb, argnum);
 			if (t->byref)
 				mono_mb_emit_byte (mb, CEE_LDIND_REF);
 			mono_mb_emit_ldloc (mb, conv_arg);
-			mono_mb_emit_icall_id (mb, conv_to_icall (MONO_MARSHAL_FREE_LPARRAY, NULL));
+
+			if (param_num == -1) {
+				mono_mb_emit_icon (mb, num_elem);
+			} else {
+				mono_mb_emit_ldarg (mb, param_num);
+				if (num_elem > 0) {
+					mono_mb_emit_icon (mb, num_elem);
+					mono_mb_emit_byte (mb, CEE_ADD);
+				}
+				mono_mb_emit_byte (mb, CEE_CONV_OVF_I);
+			}
+
+			mono_mb_emit_icall_id (mb, conv_to_icall (MONO_MARSHAL_CONV_LPARRAY_ARRAY, NULL));
```

Comment:
Why doesn't the part in `if (need_convert || need_free) {` in the case already handle the conversion? Can it be fixed to include the case you're seeing?

Reply:
Well that case seems to be for string and string builder objects:
* need_convert is only set for unicode char arrays,
* need_free is set for stringbuilder class instances.

As I see it, the `MARSHAL_ACTION_CONV_IN` side handles the marshalling of `ARRAY` to `LPARRAY` through `mono_array_to_lparray_impl`, and it seems much more natural to handle the `MARSHAL_ACTION_CONV_OUT` marshalling symmetrically in the function that was only freeing the `LPARRAY`.

Also as it's executed at runtime it's simpler than having to emit IR to do the same.

Reply:
If I'm reading this right, `need_convert` is true if there's an `OutAttribute` on the parameter (maybe it should be using `is_out` instead.

Reply:
Eh indeed, I got lost in the parentheses sorry.